src/main/java/org/eln2/mc/common/cells/foundation/CellBase.kt

```diff
@@ -0,0 +1,161 @@
+package org.eln2.mc.common.cells.foundation
+
+import mcp.mobius.waila.api.IPluginConfig
+import net.minecraft.resources.ResourceLocation
+import org.eln2.mc.common.cells.foundation.objects.ElectricalConnectionInfo
+import org.eln2.mc.common.cells.foundation.objects.ElectricalObject
+import org.eln2.mc.common.cells.foundation.objects.SimulationObjectSet
+import org.eln2.mc.common.cells.foundation.objects.SimulationObjectType
+import org.eln2.mc.common.space.RelativeRotationDirection
+import org.eln2.mc.integration.waila.IWailaProvider
+import org.eln2.mc.integration.waila.TooltipBuilder
+
+data class CellConnectionInfo(val cell: CellBase, val sourceDirection: RelativeRotationDirection)
+
+/**
+ * The cell is a physical unit, that may participate in multiple simulations. Each simulation will
+ * have a Simulation Object associated with it.
+ * Cells create connections with other cells, and objects create connections with other objects of the same simulation type.
+ * */
+abstract class CellBase(val pos: CellPos, val id: ResourceLocation) : IWailaProvider {
+    lateinit var graph: CellGraph
+    lateinit var connections: ArrayList<CellConnectionInfo>
+
+    val hasGraph get() = this::graph.isInitialized
+
+    fun removeConnection(cell: CellBase) {
+        val target: CellConnectionInfo = connections.firstOrNull { it.cell == cell }
+            ?: error("Tried to remove non-existent connection")
+
+        connections.remove(target)
+    }
+
+    var container: ICellContainer? = null
+
+    private var createdSet: SimulationObjectSet? = null
+
+    /**
+     * Called once when the object set is requested. The result is then cached.
+     * @return A new object set, with all the desired objects.
+     * */
+    abstract fun createObjectSet(): SimulationObjectSet
+
+    private val objectSet: SimulationObjectSet
+        get() {
+            if (createdSet == null) {
+                createdSet = createObjectSet()
+            }
+
+            return createdSet!!
+        }
```

Comment:
I think this might make more sense as a Kotlin list/set of simulation objects that we can extend with the functionality needed, instead of creating all simulators in the object set. Maybe I'm missing something here.